Salesforce 9.8.0 M4 - Connector Files (#481)
mulesoft/docs-connectors: modules/ROOT/pages/salesforce/salesforce-connector-98.adoc

> = Salesforce Connector - Get Started - Mule 4

Reviewer: EDIT INSTRUCTIONS:

Could I get a green only check done today before 1 pm? #481 — Please ignore the connector reference. Thank you.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

Salesforce Connector v9.8

> Anypoint Salesforce Connector (Salesforce Connector) enables you to create applications that react to Salesforce events

Reviewer: ```suggestion
Anypoint Connector for Salesforce (Salesforce Connector) enables you to create applications that react to Salesforce events
```

such as adding, changing, or deleting objects, topics, documents, and channels.

Salesforce Connector enables you to connect to the Salesforce APIs. This connector exposes methods for accessing Salesforce, including working with the https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_dev_process_chapter.htm[Salesforce Apex classes]. 

There are three ways to use a Salesforce connector in your application: as an outbound connector, an inbound connector, or a streaming inbound connector. 

This connector works with the Salesforce SOAP API, REST API, Bulk API, and Streaming API, depending on the operation you configure. Each API call uses an XML request and response over an HTTPS connection. All required request headers, error handling, and HTTPS connection configurations are built into the connector.

Release Notes: xref:release-notes::connector/salesforce-connector-release-notes-mule-4.adoc[Salesforce Connector Release Notes]

== About Connectors

> Anypoint connectors are Mule runtime engine extensions that enable you to connect to APIs and resources on external systems, such as Salesforce, ServiceNow, and Twitter.

Reviewer: ```suggestion
Anypoint Connectors are Mule runtime engine extensions that enable you to connect to APIs and resources on external systems, such as Salesforce, ServiceNow, and Twitter.
```

== Prerequisites

Before creating an app, you must have access to the Salesforce target resource and Anypoint Platform. You must also understand how to create a Mule app using Design Center or Anypoint Studio.

> This document assumes that you are familiar with Salesforce, HTTP, SOAP, REST, Bulk and Streaming APIs, Mule runtime engine (Mule), Anypoint Connectors, Anypoint Studio (Studio), Mule concepts, elements in a Mule flow, and Global Elements.

Reviewer: GLOBALLY remove "This document assumes...." It is both anthropomorphic, and it is self-referential, two things that we want to avoid.

Requirements for using the connector:

> 

Reviewer: GLOBALLY change bullet lists to follow the guidelines in our style reference, as follows:

> * https://developer.salesforce.com[Salesforce developer account].

Reviewer: ```suggestion
* https://developer.salesforce.com[Salesforce developer account]
```

> * https://help.salesforce.com/articleView?id=user_security_token.htm[Salesforce security token] - You can receive a new security token by email if you run *Reset Security Token* through *My Personal Information* from the Salesforce Setup pages.

Reviewer: ```suggestion
* https://help.salesforce.com/articleView?id=user_security_token.htm[Salesforce security token]
You can receive a new security token by email if you run *Reset Security Token* through *My Personal Information* from the Salesforce Setup pages.
```

> * Consumer key and Secret (available in your Salesforce developer account) - Required if you are using the OAuth API.

Reviewer: ```suggestion
* Consumer key and Secret (available in your Salesforce developer account)
Required if you are using the OAuth API.
```

> * Namespace and schema location: If you plan to create the XML for your Mule application, you need to include the correct namespace and schema location in your XML file. Anypoint Studio adds this information to the XML file automatically when you add the Salesforce connector to a flow in a Mule app.

Reviewer: ```suggestion
* Namespace and schema location
If you plan to create the XML for your Mule application, you must include the correct namespace and schema location in your XML file. Anypoint Studio adds this information to the XML file automatically when you add the Salesforce Connector to a flow in a Mule app.
```

> [NOTE]

Reviewer: Instead of a NOTE, I would call the first two restrictions and perhaps add the third to the requirements text.
Avoid using "Note" as a dumping ground for stuff that does not gracefully fit elsewhere; there is almost always a better alternative.
In fact, the first two might be mentioned in the main narrative, when we name the other Salesforce APIs.

====

* The Salesforce connector does not expose all possible operations of the Salesforce APIs.
* The Salesforce connector does not provide access to the Chatter API or the Tooling API.
* See http://blogs.developerforce.com/tech-pubs/2011/10/salesforce-apis-what-they-are-when-to-use-them.html[Salesforce APIs: What they are and when to use them] and https://help.salesforce.com/HTViewHelpDoc?id=integrate_what_is_api.htm[Which API should I use?].
====

== Audience

> 

Reviewer: GLOBAL
I have marked this many times.
Please update the template and this text to use the list style described in our style reference and exemplified in the list-item edits above.

* Starting users -- Read Get Started (this document) and create your Mule app in 
xref:salesforce/salesforce-connector-design-center.adoc[Design Center] or 
xref:salesforce/salesforce-connector-studio.adoc[Anypoint Studio]. 
The xref:salesforce/salesforce-connector-config-topics.adoc[Additional Configuration] topic
helps you extend your understanding of the Salesforce connector.
* Power users - Read xref:salesforce/salesforce-connector-xml-maven.adoc[XML and Maven Support], 
the xref:salesforce/salesforce-connector-config-topics.adoc[Additional Configuration] topic, 
and xref:salesforce/salesforce-connector-examples.adoc[Examples].

=== Subject Overview

> 

Reviewer: Is this, more specifically, an overview to documentation titles and their content? 
Doesn't the TOC handle this; do we really need this table?

[%header%autowidth.spread]
|===

> |Subject |Description

Reviewer: It would be nice if the description texts were parallel: all fragments or all sentences; all imperative (do this) or all noun based (this is what this is).

|xref:salesforce/salesforce-connector-examples.adoc#accandtrans[Accept and Transform Data] |Example that accepts and transforms data from Salesforce.

> |xref:salesforce/salesforce-connector-studio.adoc#apexsets[Apex SOAP and REST] |Configure for APEX SOAP API and APEX REST API

Reviewer: Does not end in period, like the others.

|<<apis,APIs>> |Salesforce APIs that this connector can access.
|<<developapp, Application Development>> |Processing order and integration patterns.
|xref:salesforce/salesforce-connector-config-topics.adoc#batchdata[Batched Data] |Batching data using the Salesforce Bulk API.
|xref:salesforce/salesforce-connector-config-topics.adoc#customevents[Custom Event Notifications] |Get custom event notifications over a streaming channel.
|xref:salesforce/salesforce-connector-config-topics.adoc#eventsandtopics[Events] |Receive Salesforce events by subscribing to a topic.
|xref:salesforce/salesforce-connector-config-topics.adoc#leadconvert[Lead Convert Request ID] |Specify a lead ID using DataWeave.
|xref:salesforce/salesforce-connector-examples.adoc#parentchild[Parent Child Relationships] |Example for creating and updating parent and child relationships.
|xref:salesforce/salesforce-connector-config-topics.adoc#streamsub[Streaming Channel] |Subscribe to a Salesforce streaming channel.

> |<<Anypoint Exchange Templates and Examples>> |Simplify Salesforce connector development with a template or an example.

Reviewer: ```suggestion
|<<Anypoint Exchange Templates and Examples>> |Simplify Salesforce Connector development with a template or an example.
```

|xref:salesforce/salesforce-connector-config-topics.adoc#eventsandtopics[Topics] |Subscribe to a Salesforce topic.
|===

> The Anypoint Salesforce Connector lets you connect to the Salesforce APIs. This connector exposes methods for accessing Salesforce, including working with the https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_dev_process_chapter.htm[Salesforce Apex classes]. There are three ways to use a Salesforce connector in your application: as an outbound connector, an inbound connector, or a streaming inbound connector. 

Reviewer: ```suggestion
Salesforce Connector enables you to connect to the Salesforce APIs. This connector exposes methods for accessing Salesforce, including working with the https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_dev_process_chapter.htm[Salesforce Apex classes]. There are three ways to use a Salesforce connector in your application: as an outbound connector, an inbound connector, or a streaming inbound connector. 
```

Reviewer: Did you intend to repeat this text?

[[developapp]]
== Develop an Application

Build an application in this order:

. Configure the connector.
. Test the connection.
. Build the rest of your flow.
. Add and configure DataWeave.

> When developing an app, the Salesforce connector provides these integration patterns

Reviewer: ```suggestion
When developing an app, Salesforce Connector provides these integration patterns
```

you can use with the Salesforce APIs:

> 

Reviewer: Please reformat, as indicated above.

* Batch Data Synchronization - An external system accesses, changes, deletes, or adds data in Salesforce in batches, and vice versa (Salesforce to external system).

> * Remote Call-In - An external system accesses, changes, deletes or adds data in Salesforce, and vice versa (Salesforce to external system).

Reviewer: missing comma
```suggestion
* Remote Call-In - An external system accesses, changes, deletes, or adds data in Salesforce, and vice versa (Salesforce to external system).
```

* Fire and Forget Remote Process Invocation - Salesforce initiates a process in a third-party system and receives an acknowledgment that the process has started. The third-party system continues processing independent of Salesforce.

> * Request-Reply Remote Process Invocation - Salesforce kicks off a process in a remote system, waits for the remote system to finish processing, then accepts control back again from the remote system.

Reviewer: Replace "kicks off" with "initiates," as above.

Reviewer: ", then" > ", and then"

GLOBAL: "then" alone is not a conjunction.

* User Interface Update Based on Data Changes - The Salesforce UI updates in response to a change in a third-party system.

[[apis]]
== APIs for the Salesforce Connector

> The Salesforce connector works with the Salesforce APIs in the sections that follow. The Salesforce connector does not provide access to the Chatter API or the Tooling API. Each API call uses an XML request and response over an HTTPS connection. All required request headers, error handling, and HTTPS connection configurations are built into the connector.

Reviewer: ```suggestion
Salesforce Connector works with the Salesforce SOAP API, REST API, Bulk API, and Streaming API, but does not provide access to the Chatter API or the Tooling API. Each API call uses an XML request and response over an HTTPS connection. All required request headers, error handling, and HTTPS connection configurations are built into the connector.
```

=== Apex SOAP API

> 

Reviewer: GLOBALLY lowercase "web" unless part of a proper noun.

The Salesforce https://www.salesforce.com/us/developer/docs/apexcode/[Apex SOAP API] exposes Apex class methods as custom SOAP Web service calls. This allows an external application to invoke an Apex Web service to perform an action in Salesforce.

=== Apex REST API

> The Salesforce https://developer.salesforce.com/page/Creating_REST_APIs_using_Apex_REST[Apex REST API] creates your own REST-based web services using Apex. It has all of the advantages of the REST architecture, and provides the ability to define custom logic and includes automatic argument/object mapping.

Reviewer: ```suggestion
The Salesforce https://developer.salesforce.com/page/Creating_REST_APIs_using_Apex_REST[Apex REST API] creates your own REST-based web services using Apex. It has all of the advantages of the REST architecture while adding the ability to define custom logic and including automatic argument-object mapping.
```

Reviewer: Just curious: do we typically document Salesforce functionality? I mean, it's kind of a third-party here, right?

=== Bulk API

> The Salesforce https://www.salesforce.com/us/developer/docs/api_asynch/[Bulk API] quickly and securely load batches of your organization's data into Salesforce. See also: xref:salesforce/salesforce-connector-config-topics.adoc#batchdata[Load Data in Batches].

Reviewer: ```suggestion
The Salesforce https://www.salesforce.com/us/developer/docs/api_asynch/[Bulk API] quickly and securely loads batches of your organization's data into Salesforce. See xref:salesforce/salesforce-connector-config-topics.adoc#batchdata[Load Data in Batches].
```

=== Metadata API

The Salesforce http://www.salesforce.com/us/developer/docs/api_meta/[Metadata API] manages customizations and build tools that work with the metadata model, not the data itself.

=== SOAP API

> The Salesforce http://www.salesforce.com/us/developer/docs/api/index.htm[SOAP API] provides secure access to your organization's information on Salesforce. Most of the operations that the Salesforce connector performs map to this API.

Reviewer: ```suggestion
The Salesforce http://www.salesforce.com/us/developer/docs/api/index.htm[SOAP API] provides secure access to your organization's information on Salesforce. Most of the operations that Salesforce Connector performs map to this API.
```

> All the Salesforce operations performed through the SOAP API have an optional parameter called "Headers" that can take any of the following https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/soap_headers.htm[Salesforce SOAP Headers]:

Reviewer: Is it clear to say that the optional parameter Headers takes Salesforce SOAP Headers? Sounds a little convoluted to me.

Also, I'm not sure I understand why there is a colon leading to an example that is just a progression of header names. This is no easier to read than it would be in the original sentence. Did you want to use a bullet list?

Remove the period.

AllOrNoneHeader, AllowFieldTruncationHeader, AssignmentRuleHeader, CallOptions,
EmailHeader, LocaleOptions, MruHeader, OwnerChangeOptions, QueryOptions,
UserTerritoryDeleteHeader, DuplicateRuleHeader.

=== Streaming API

> The Salesforce http://www.salesforce.com/us/developer/docs/api_streaming/[Streaming API] securely receives notifications for changes to your organization's information in Salesforce. See xref:salesforce/salesforce-connector-config-topics.adoc#receivedata[Receive Inbound Data From Salesforce] for more information on the use of the Streaming API.

Reviewer: information ABOUT

== Anypoint Exchange Templates and Examples

https://www.mulesoft.com/exchange/[Anypoint Exchange] provides templates
you can use as a starting point for your app, as well as examples that illustrate a complete solution.

=== Templates in Exchange

https://www.mulesoft.com/exchange/?search=salesforce&type=template[Anypoint Exchange templates] provide starting points for Anypoint Studio projects such as:

> * Account Broadcast template: https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc.wday.sap.db-account-broadcast/[Salesforce to Salesforce, Workday, SAP, and Database]

Reviewer: ```suggestion
* Account Broadcast template
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc.wday.sap.db-account-broadcast/[Salesforce to Salesforce, Workday, SAP, and Database]
```

Reviewer: GLOBALLY remove the colons at the end of the bullet-list items.

* Aggregation:

> https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc-contact-aggregation/[contact]

Reviewer: Missing comma...although I would not punctuate any of these as if they were sentences, without providing a subject and verb....

https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc-opportunity-bidirectional-sync/[opportunities], 
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc-user-bidirectional-sync/[users]
** Other sources: 
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2db-account-bidirectional-sync/[Database], 
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2msdyn-account-bidirectional-sync/[Microsoft Dynamics CRM], 
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sieb-account-bidirectional-sync/[Siebel]
* Migration: https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sap-account-migration/[accounts],
https://www.mulesoft.com/exchange/org.mule.templates/template-sap2sfdc-contact-migration/[contacts],
https://www.mulesoft.com/exchange/org.mule.templates/template-sfdc2sfdc-opportunity-migration/[opportunities]

=== Examples in Exchange

https://www.mulesoft.com/exchange/?search=salesforce&type=example[Anypoint Exchange examples] enable you to create complete Anypoint Studio projects such as:

> * Batch processing: https://www.mulesoft.com/exchange/org.mule.examples/salesforce-to-MySQL-DB-using-Batch-Processing/[database], 

Reviewer: ```suggestion
* Batch processing
https://www.mulesoft.com/exchange/org.mule.examples/salesforce-to-MySQL-DB-using-Batch-Processing/[database], 
```

https://www.mulesoft.com/exchange/org.mule.examples/import-leads-into-salesforce/[import leads]
* https://www.mulesoft.com/exchange/org.mule.examples/cache-scope-with-salesforce-contacts/[Contacts]
* https://www.mulesoft.com/exchange/org.mule.examples/salesforce-data-retrieval/[Data retrieval]
* https://www.mulesoft.com/exchange/org.mule.examples/dataweave-with-flowreflookup/[DataWeave]

> == Next

Reviewer: Is this consistent with the template?
